### PR TITLE
Closes #911: Implement `find_locations` and `findall` on regexes for `Strings`

### LINKED ISSUE #911: Implement `find_locations` and `findall` on regexes for `Strings`
In support of #894 :
- Implement `Strings.find(pattern)` which returns the number of matches, start positions, and lengths of matches (return value similar to a list of [`reMatch` records ](https://chapel-lang.org/docs/modules/standard/Regexp.html?highlight=regexp#Regexp.reMatch))
- Implement `Strings.slice` which uses the return values of `find` and returns the slices of `Strings` that match `pattern`

### PR DESCRIPTION
**Pull request by pierce314159**
This PR (closes #911):
- Adds `find_locations`/`findall` to `Strings` and unit tests for this functionality. Also caches previous `Strings.find_locations` results
- `find_locations(pattern)` 
  - Finds pattern matches and returns pdarrays containing the number, start postitions, and lengths of matches
  - Results of `find_locations` are cached in `regex_dict`
- `findall(pattern, return_match_origins)`
  - Return all non-overlapping matches of pattern in Strings as a new Strings object. If `return_match_origins` is True, return a pdarray containing the index of the original string each pattern match is from
  - Uses the output of `find_locations`

Example:
```python
>>> strings = ak.array(['', '____', '_1_2', '3___4___', '5'])
>>> strings.find('_+')
(array([0 1 2 2 0]), array([1 6 8 12 16]), array([4 1 1 3 3]))

>>> strings.print_cached_regex_patterns()
dict_keys(['_+'])

>>> strings.slice('_+', return_match_origins=True)
(array(['____', '_', '_', '___', '___']), array([1 2 2 3 3]))

>>> strings.slice('\\d', return_match_origins=True)
(array(['1', '2', '3', '4', '5']), array([2 2 3 3 4]))

>>> strings.print_cached_regex_patterns()
dict_keys(['_+', '\\d'])

>>> strings.register('myStr')
array(['', '____', '_1_2', '3___4___', '5'])

>>> ak.clear()
>>> strings.cached_regex_patterns()
dict_keys([])

>>> strings.slice('\\d')
array(['1', '2', '3', '4', '5'])

>>> strings.print_cached_regex_patterns()
dict_keys(['\\d'])
```

### CODE REVIEW

**Review by reuster986 — COMMENTED** [review]
Great work! I like the API here. Just a couple of questions:

1) I wonder if we should change the names of the `Strings.slice` method to `Strings.search`. I'm envisioning a future `slice` method that pulls out substrings based on arrays of start and end indices. To me, a function that takes a regex and returns substrings feels more like a `search` than a `slice`. What do you think?

2) For both @pierce314159 and @ronawho what is the relationship of this PR to #928 and #931? Does the code in this PR need to change based on those other PRs, or is it abstracted enough to merge now?

**Comment by pierce314159** [comment]
> 
> 
> Great work! I like the API here. Just a couple of questions:
> 
>     1. I wonder if we should change the names of the `Strings.slice` method to `Strings.search`. I'm envisioning a future `slice` method that pulls out substrings based on arrays of start and end indices. To me, a function that takes a regex and returns substrings feels more like a `search` than a `slice`. What do you think?
> 
>     2. For both @pierce314159 and @ronawho what is the relationship of this PR to #928 and #931? Does the code in this PR need to change based on those other PRs, or is it abstracted enough to merge now?

Thanks! 
1. I agree `Strings.search` feels like a better fit and leaves the `slice` name available
2. I believe the only thing that should need to change is in `findMatches`
```C
var matches = myRegex.matches(interpretAsString(origVals[off..#len])); 
->
var matches = myRegex.matches(interpretAsBytes(origVals, off..#len, borrow=true));
```

File: arkouda/strings.py [hunk 1]
```diff
@@ -255,6 +260,146 @@ def get_lengths(self) -> pdarray:
...
+        """
+        sym_tab = list_symbol_table()
+        self.regex_dict = {key: val for key, val in self.regex_dict.items() if all([pda.name in sym_tab for pda in val])}
+
+    def print_cached_regex_patterns(self):
+        """
+        Prints the regex patterns for which Strings.find(pattern) has been cached
+        """
+        self._refresh_regex_dict()
+        print(self.regex_dict.keys())
```
Comment (glitch) [inline arkouda/strings.py:275]: It looks like `_refresh_regex_dict` is only called by `print_cached_regex_patterns`, I'm wondering if a better approach would be to return a copy of the dictionary you made in `_refresh_regex_dict` if you're worried about a user modifying the state of it.  Then the user can choose to print it or not... having a function that prints feels weird to me.

Comment (pierce314159) [reply arkouda/strings.py:275]: ooo yeah i agree! that makes more sense

**Comment by ronawho** [comment]
> For both @pierce314159 and @ronawho what is the relationship of this PR to #928 and #931? Does the code in this PR need to change based on those other PRs, or is it abstracted enough to merge now?

`interpretAsString` will need to be updated to pass `borrow=true` to borrow the segmented array data instead of allocating a new string. I can roll that into #931

File: arkouda/strings.py [hunk 2]
```diff
@@ -255,6 +260,146 @@ def get_lengths(self) -> pdarray:
...
+                                        self.offsets.name,
+                                        self.bytes.name,
+                                        json.dumps([pattern]))
+            repMsg = cast(str, generic_msg(cmd=cmd, args=args))
+            arrays = repMsg.split('+', maxsplit=2)
+            self.regex_dict[pattern] = (create_pdarray(arrays[0]), create_pdarray(arrays[1]), create_pdarray(arrays[2]))
+        return self.regex_dict[pattern]
+
+    @typechecked
+    def slice(self, pattern: Union[bytes, str_scalars], return_match_origins: bool = False) -> Union[Strings, Tuple]:
```
Comment (glitch) [inline arkouda/strings.py:343]: I agree with @reuster986 on the naming here.  I'd try to make the naming match the `re2` naming of what's happening here if possible.  `slice` seems to imply a different context..something you apply to the structure of an array/list.  Looking at the ongoing conversation it sounds like you are already going to change it.

**Comment by pierce314159** [comment]
@reuster986 and @glitch, looking at the python `re` api (which seems like what we should try to mimic to me, let me know if anyone thinks otherwise), it seems `slice` here is essentially [`findall`](https://docs.python.org/3/library/re.html#re.findall). So I think I'll rename `slice` to that. I think I'll also change `find` to `find_locations`

Let me know if there's any issue with this!

**Comment by reuster986** [comment]
@pierce314159 `findall` and `find_locations` sound good to me.

**Review by glitch — APPROVED** [review]

**Review by reuster986 — APPROVED** [review]

**Pull request merged**